src/autoscaler/scalingengine/server/applock_test.go

```diff
@@ -0,0 +1,106 @@
+package server_test
+
+import (
+	. "autoscaler/scalingengine/server"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"time"
+)
+
+var (
+	ch      chan string
+	appLock *AppLock
+	err     interface{}
+)
+
+var _ = Describe("Applock", func() {
+	BeforeEach(func() {
+		appLock = NewAppLock()
```

Comment:
I see problems with AppLock.
1. At runtime, we keep adding a lock for each AppId but I never see where we drop all the locks, i.e., do any cleanup.

2. As far as the test cases are concerned, the problem you have is that you have a goroutine using the AppLock which uses it after the next test starts. Your afterEach only guarantees that the goroutine will do something, but it doesn't guarantee that it happens before the next test.
You can either pass the applock directly to the goroutine, or wait for the goroutine to exit properly before allowing the next test to run.